### DISCUSSION TITLE
How to handle secrets during docker build stage in CI ?

### DISCUSSION THREAD

**Discussion by ohaag** [symfony/symfony#45975] [Q&A] [answered]
Hello,

I tried several time to discuss this topic on support slack without any answer so I try here:

I'm using symfony secrets and here is the issue I encounter:
I store the private key into my company vault solution.
During the CI build of the docker artifact, I tried to build without that private key but auto-scripts after composer install (cache:clear & assets:install) are requiring env variables like DATABASE_URL (probably to build the container and some DI in it).

I can't composer install with --no-scripts options because assets:install is needed for the next step of the docker build : yarn install

So I ended to put that private key into my CI during the build to remove it at the end and add it again during deployment.

Another weird thing is that I build my App in APP_ENV=prod even if I will deploy it on another env at the end because it does not matter during build state.

Is it the normal way ? Another way would be to build one artifact for each env but I don't think it is a good solution.

In the same topic, where should we composer dump-env ? I went for executing it on deploy right before running php-fpm, meaning I have to download composer just for that and remove it after.

**Comment by MichaelBrauner** [comment]
As long as you don't use the [symfony binary](https://github.com/symfony-cli/symfony-cli) you need to handle exposing your ENV vars to your docker environment manually (or you read it inside your CI script from .env with `source .env`). 

Regarding the problem with the .env file and the production mode: https://symfony.com/doc/current/configuration.html#overriding-environment-values-via-env-local

**Reply by ohaag** [reply]
I don't understand what the symfony binary would be used for in this case.

The fact is that these are not env vars. They are secrets. If you declare a secret key in the .env, the .env will be the final value. So you can't declare any secret template in your .env or it will override the real secret.
(btw, that's another thing I think it's a loss: you can't have a default env template with SECRET=CHANGE_THIS as you could do with .env or parameters.yml.dist before)

So how am I supposed to expose them to docker ?
In order to keep my real .env file, I guess I could : 
 - write all secret keys into .env.local (from any file I would have prepared, let's say : secret_keys.yaml)
 - docker build
 - remove .env.local

It implies obscure practice as maintaining that keys list into a random secret_keys.yaml.
I'd rather give it access to the secret vault (aka the private key) and then remove it. (nothing to maintain)

Shouldn't the secret keys be known as env var in Symfony even if it has no key to decrypt them ? It would make disapear the issue about existence check. It would not bypass advance synthax control I guess, as variables would be empty but should there be some?
(I'm not sure why, but it seems my project needs a valid DATABASE_URL to work. As if it'd make a query at container build; maybe a bad DI handling ?)

---

For the last link, it shows how to override .env. I've no issue with that, I override them with native env variables (from a K8S configmap). What I consider weird is the fact I have to build the docker artifact with wathever APP_ENV because I can't build without one; even though it makes no sense as only the deployment will set the final env var.

---

A trace of the CI/CD process I made:

- Pipeline:
  - Pull of decrypt.private_key
  - docker build :
    - composer install (works only because symfony can decrypt needed secrets env vars)
    - remove decrypt.private_key
  - push artifact on registry
  - triggers other pipeline deploying on kubernetes :
    - pull decrypt.private_key
    - helm upgrade ...
      - private key is mounted as a "volume" into K8S deployment.
      - composer dump-env is now possible with an init-container

Without symfony secrets, I would just fetch all my secrets before deployment to provide them to K8S and I wouldn't need them in docker as they would be in .env.
But as said, now I can't put the keys in .env as it would override my secrets.

**Reply by MichaelBrauner** [reply]
The symfony binary exposes env vars: https://symfonycasts.com/screencast/symfony-doctrine/docker-env-vars.

For the rest of your question: I don't use kubernetics.

I deploy my small apps with Laravel Forge on a Hetzner Server (very handy). 
For bigger solutions I go with symfony clould (platform.sh).

**Comment by GromNaN** [comment, marked as answer]
You cannot use the same cache for `dev` and `prod`, `APP_ENV` alters the structure of the project (think enabled bundles). But you could warmup both caches in your docker container.

For secret variables, I set default values in `.env`. So that the symfony cache can be built even if the secrets are not available.

Env vars must not be used during container build, if their value can be updated at runtime: see https://github.com/symfony/symfony/issues/40794

I wrote an article about Vault with Symfony: https://dev.to/gromnan/store-secrets-in-vault-with-symfony-51ai

**Reply by ohaag** [reply]
If you define you default value for Secrets in you .env, how do you read the real values after deployment ? Even with secret:decrypt-to-local, it will not put your secret in .env.prod.local if the key is already defined in .env.

-----

I didn't care too much about warming the cache upfront but is it a recurrent practice: building cache for each env in docker build than can be deployed ?

On the contrary direction, I went from Vault to Symfony secrets in order to have everything closer to the dev, but I understand vault pros, especially with live considerations.

Considering "Use env vars ONLY if you want to be able to change values WITHOUT redeploying",  should DATABASE_URL be in parameters? How do you mix parameters.yml with SECRET. I'm quite lost on the guidelines sorry ^^.
